pkg/component/starter_project.go

```diff
@@ -0,0 +1,215 @@
+package component
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	devfilev1 "github.com/devfile/api/pkg/apis/workspaces/v1alpha2"
+	parsercommon "github.com/devfile/library/pkg/devfile/parser/data/v2/common"
+
+	"github.com/devfile/library/pkg/devfile/parser"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport/http"
+	"github.com/openshift/odo/pkg/log"
+	registryUtil "github.com/openshift/odo/pkg/odo/cli/registry/util"
+	"github.com/openshift/odo/pkg/util"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	devFile                   = "devfile.yaml"
+	defaultStarterProjectName = "devfile-starter-project-name"
+)
+
+var DevfilePath = filepath.Join("./", devFile)
+
+func checkoutProject(subDir, zipURL, path, starterToken string) error {
+
+	if subDir == "" {
+		subDir = "/"
+	}
+	err := util.GetAndExtractZip(zipURL, path, subDir, starterToken)
+	if err != nil {
+		return errors.Wrap(err, "failed to download and extract project zip folder")
+	}
+	return nil
+}
+
+// GetStarterProject gets starter project value from flag --starter.
+func GetStarterProject(projects []devfilev1.StarterProject, projectPassed string) (project *devfilev1.StarterProject, err error) {
+
+	nOfProjects := len(projects)
+
+	if nOfProjects == 0 {
+		return nil, errors.Errorf("no starter project found in devfile.")
+	}
+
+	// Determine what project to be used
+	if nOfProjects == 1 && projectPassed == defaultStarterProjectName {
+		project = &projects[0]
+	} else if nOfProjects > 1 && projectPassed == defaultStarterProjectName {
+		project = &projects[0]
+		log.Warning("There are multiple projects in this devfile but none have been specified in --starter. Downloading the first: " + project.Name)
+	} else { //If the user has specified a project
+		projectFound := false
+		for indexOfProject, projectInfo := range projects {
+			if projectInfo.Name == projectPassed { //Get the index
+				project = &projects[indexOfProject]
+				projectFound = true
+				break
+			}
+		}
+
+		if !projectFound {
+			return nil, errors.Errorf("the project: %s specified in --starter does not exist", projectPassed)
+		}
+	}
+
+	return project, err
+
+}
+
+// Downloads first starter project from list of starter projects in devfile
+// Currently type git with a non github url is not supported
+func DownloadStarterProject(devObj parser.DevfileObj, starterProject *devfilev1.StarterProject, decryptedToken string, registryName string) error {
```

Comment:
I am sensing that this function could be part of `devfilev1.StarterProject` struct as a method, but it then needs to land in the api repo